### ISSUE TITLE
About Thounderbolt 3 port

### ISSUE THREAD

**Issue by qianwuyu** [gorquan/OC-XPS-7590#7]
您好，雷电的解决方案我想在黑果小兵所做的适配里有相关的SSDT，因为截图中显示出了雷电的设备信息。推测该EFI应该是黑果小兵为某人定制机型EFI的时候所做，没有后续更新，但[SSDT可参考](https://github.com/daliansky/XPS15-7590-Hackintosh)。
![雷电信息](https://github.com/daliansky/XPS15-7590-Hackintosh/blob/master/screenshot/2.png?raw=true)

**Comment by gorquan** [comment]
@qianwuyu 雷电我暂时没有什么设备可以测试，所以就先搁置一边了，谢谢你的提议！～

**Comment by CryptoNeverSleeps** [comment]
Commenting on my Hackintosh....I added the TB3 and USB C .aml files from Daliansky Github... 

<img width="1920" alt="Screen Shot 2020-08-26 at 11 25 43 AM" src="https://user-images.githubusercontent.com/36324479/91323744-3bedc800-e78f-11ea-84a2-efaa1c27154a.png">

**Comment by CryptoNeverSleeps** [comment]
I'll need to purchase a TB3 Device to test the TB3 Port Functionality... The port has always worked for Type-C Devices... including Dell Type-C docking station.

**Comment by qianwuyu** [comment]
> I'll need to purchase a TB3 Device to test the TB3 Port Functionality... The port has always worked for Type-C Devices... including Dell Type-C docking station.

**Comment by qianwuyu** [comment]
@CryptoNeverSleeps :
Nice👍! Waiting for news. And you mean it can plug in/out with Type-C Devices anytimes?

**Comment by CryptoNeverSleeps** [comment]
Yes, Even without the ACPI... I could always plug in/out with Type-C devices anytime... I tested yesterday with my iPad Pro 12.9" 3rd Generation...   Not sure when I'll purchase a TB3 Capable device... Type-C has always worked... I've actually booted macOS Catalina using the Type-C / Thunderbolt Port.... I'm using a NVMe Enclosure......

**Comment by qianwuyu** [comment]
> Yes, Even without the ACPI... I could always plug in/out with Type-C devices anytime... I tested yesterday with my iPad Pro 12.9" 3rd Generation... Not sure when I'll purchase a TB3 Capable device... Type-C has always worked... I've actually booted macOS Catalina using the Type-C / Thunderbolt Port.... I'm using a NVMe Enclosure......

Sadly, I did it like you(changed the .aml files), but there's nothing changed, I even can't make it  useful if  plug in at start my machine. Or the neweast version(from this repository) can't make that? 

I'm sorry to bother you after so many days, but really hope you can answer me, or add a easy contact ifomation? Yeah, we can exchange E-Maill! My E-Mail address is: ZekShawn@smail.nju.edu.cn / Eternal@hnu.edu.cn. Thank you for your reply.

**Comment by CryptoNeverSleeps** [comment]
Can you provide screenshots?   Explain further the problem?

**Comment by CryptoNeverSleeps** [comment]
https://osy.gitbook.io/hac-mini-guide/details/thunderbolt-3-fix/ -  If you're using thunderbolt.... you still need to fix thunderbolt... The modified ACPI for TB3 only allows Type-C devices until you fix Thunderbolt according to this link...

**Comment by CryptoNeverSleeps** [comment]
Another Guide about TB3 - https://dortania.github.io/OpenCore-Post-Install/universal/sleep.html#fixing-thunderbolt

**Comment by qianwuyu** [comment]
> Another Guide about TB3 - https://dortania.github.io/OpenCore-Post-Install/universal/sleep.html#fixing-thunderbolt

Thx, I'll try like the html page said later because I breaked out the macOS yet. When I plug in Type-C device firstly after boot macOS, it works! But not working after that. And I am a graduate student in China, if lack of Type-C port, this machine seems not so great for macOS. Anyway, Thx! I'll try if there's spare time.

**Comment by gorquan** [comment]
@CryptoNeverSleeps @qianwuyu ThunderBolt related DSDT files have been added to Opencore, but I don’t have a device and there is no way to test it. After adding ThunderBolt, my battery life started to drop, I hope it is not caused by ThunderBolt. If you have a corresponding ThunderBolt device, you can help me test it, thank you very much!

**Comment by gorquan** [comment]
@CryptoNeverSleeps @qianwuyu In subsequent versions, if ThunderBolt is not effective, I will remove ThunderBolt from Opencore

**Comment by qianwuyu** [comment]
@gorquan Great work & thank you. But I bumped into a strange thing that you have not been mentioned in README. I can't make Type-C port work after the first pluging in. Really bad experience. Of course, it maybe just occur on my machine. My machine is some different from you, there's Intel i5-9300H core inside and no any external graphic card. I don't know if that only occur after I remove the 'SSDT-NDGP.aml' in 'config.plist' file. I have attempted to add back, but it not work again, really saddly. Will you consider making release version in the future? Or add a UMA only version?

**Comment by CryptoNeverSleeps** [comment]
Are you using XPS 7590? If not, what's the model of your laptop?

**Comment by qianwuyu** [comment]
> Are you using XPS 7590? If not, what's the model of your laptop?

Sure XPS 15 7590. And I only have normal Type-C device tested. No TB3 device test environment.

**Comment by CryptoNeverSleeps** [comment]
What TYPE-C device are you using? Have you tried multiple TYPE-C devices ?

**Comment by qianwuyu** [comment]
Add my hardware information picture by ludashi. And haven't your machine turned out panic on Windows 10 after you replace the wireless model "DW1560"? Or you haven't care on Windows, haha. Don't fell panic, it works great on macOS. And I attempt change my Windows 10 versions like 1909/2004/LTSC, not effective yet. The hardware drivers were downloaded all from 'dell.com'.  
![](https://pic.downk.cc/item/5f5d7edf160a154a6776a401.png)

**Comment by qianwuyu** [comment]
> What TYPE-C device are you using? Have you tried multiple TYPE-C devices ?

Great question! The testing device is Type-C Mobile SSD. I checked the hardware info in macOS, when it pluged out. The PCI info disappeared quickly. Maybe you have some easy connect method like QQ/Wechat? My QQ number: 1732484838. My wechat code: Eternal_QXZ.

**Comment by gorquan** [comment]
> @gorquan Great work & thank you. But I bumped into a strange thing that you have not been mentioned in README. I can't make Type-C port work after the first pluging in. Really bad experience. Of course, it maybe just occur on my machine. My machine is some different from you, there's Intel i5-9300H core inside and no any external graphic card. I don't know if that only occur after I remove the 'SSDT-NDGP.aml' in 'config.plist' file. I have attempted to add back, but it not work again, really saddly. Will you consider making release version in the future? Or add a UMA only version?

你可以尝试一下XPS 9570的Type-C的补丁，9570 Type-C出现的情况与你的差不多

**Comment by qianwuyu** [comment]
@gorquan 用中文交流快多了，哈哈。9570跟7590的控制器是同一个吗？我晚上试试再装一次macOS 10.15.6，另外真的很想知道你是怎么解决的DW1560在Windows下死机的啊，我每次都是用到一定程度就死机了。。。超级无语啊。。百度了一下好像没有碰到一样的情况，查看设备信息也确实是1560。现在要用Windows就必须拆后盖换回ax200，超级不方便。

**Comment by gorquan** [comment]
> @gorquan 用中文交流快多了，哈哈。9570跟7590的控制器是同一个吗？我晚上试试再装一次macOS 10.15.6，另外真的很想知道你是怎么解决的DW1560在Windows下死机的啊，我每次都是用到一定程度就死机了。。。超级无语啊。。百度了一下好像没有碰到一样的情况，查看设备信息也确实是1560。现在要用Windows就必须拆后盖换回ax200，超级不方便。

我换上去一切正常，不过你这种情况我朋友的小新也出现过了，最后排查了一轮还是找不到解决的办法
等什么时候能再次拿到电脑了，我再排查一下

**Comment by gorquan** [comment]
> @gorquan 用中文交流快多了，哈哈。9570跟7590的控制器是同一个吗？我晚上试试再装一次macOS 10.15.6，另外真的很想知道你是怎么解决的DW1560在Windows下死机的啊，我每次都是用到一定程度就死机了。。。超级无语啊。。百度了一下好像没有碰到一样的情况，查看设备信息也确实是1560。现在要用Windows就必须拆后盖换回ax200，超级不方便。

1820A咧，一样出现问题？

**Comment by gorquan** [comment]
> @gorquan 用中文交流快多了，哈哈。9570跟7590的控制器是同一个吗？我晚上试试再装一次macOS 10.15.6，另外真的很想知道你是怎么解决的DW1560在Windows下死机的啊，我每次都是用到一定程度就死机了。。。超级无语啊。。百度了一下好像没有碰到一样的情况，查看设备信息也确实是1560。现在要用Windows就必须拆后盖换回ax200，超级不方便。

有一些7590没解决的问题我都是拿9570的配置来改，像小兵的clover 7590不能外接，我拿9570的改了正常了。所以包括现在的opencore我都有很多借鉴9570的配置去改的

**Comment by qianwuyu** [comment]
@gorquan 我没有这张卡，不过可以试试，就是折腾的来回成本有点高。希望这个机子能尽快到达可用的地步，这太折腾了。对了，那台小新最后是换1820A解决了是吗？

**Comment by gorquan** [comment]
> @gorquan 我没有这张卡，不过可以试试，就是折腾的来回成本有点高。希望这个机子能尽快到达可用的地步，这太折腾了。对了，那台小新最后是换1820A解决了是吗？

并没有，妹子最后放弃黑果了，用了ax200

**Comment by qianwuyu** [comment]
> > @gorquan 我没有这张卡，不过可以试试，就是折腾的来回成本有点高。希望这个机子能尽快到达可用的地步，这太折腾了。对了，那台小新最后是换1820A解决了是吗？
> 
> 并没有，妹子最后放弃黑果了，用了ax200

哈哈，真实。我现在也用回ax200，打算暂时放弃黑果了。黑果子老是挠心窝子，总痒，一有时间就免不了去折腾下，有时间的时候倒是可。无论如何，谢谢啦。

**Comment by CryptoNeverSleeps** [comment]
![image](https://user-images.githubusercontent.com/36324479/93262735-b9936b00-f772-11ea-9004-5f1c8fa700ad.jpeg)

![image](https://user-images.githubusercontent.com/36324479/93262852-e778af80-f772-11ea-8919-ea433b30e0c0.jpeg)


I run macOS Catalina from External Type-C NVMe Enclosure... Works without issues. ( Minus - Black screen when waking from sleep) - 

Type-C has always worked for me

XPS 7590 OLED MODEL
i7-9750H
64GB RAM
DW1820A - Wi-Fi / Bluetooth -
Pins Masked with Kapton Tape.
AirDrop, Apple Watch Unlock, Etc

**Comment by qianwuyu** [comment]
@CryptoNeverSleeps Hey, I don't think macOS can wake up from external storage devices normally. Jus for reducing the battery cost, we choose to cut down the power for USB ports when macOS comes to sleep. And I think macOS need to wake up from somewhere first, then provide power-out for external devices. Maybe not yet, it doesn't effect the issue here. What I really want to know another thing is your Windows works great as you haven't changed the wireless card before? Thanks for your reply!

**Comment by CryptoNeverSleeps** [comment]
You are incorrect with your statement about Mac's waking from external drives... This wake from sleep issue is only affecting XPS-7590 OLED Models, I had the 1080P Model and everything working 100%, Internal NVMe and External NVMe Enclosure.  I've tested the same process with OLED (Internal and External NVMe) Multiple Installs, Multiple Catalina Versions, Multiple Open Core Versions. same results as have others on Reddit..... 

Wi-Fi works perfect in Windows with DW1820A using latest Wi-Fi and Bluetooth driver from Dell.

**Comment by qianwuyu** [comment]
@CryptoNeverSleeps Learned from that, and I'll try to install macOS on my type-C device.

**Comment by qianwuyu** [comment]
好消息，我在重新安装我的macOS，当成主力用了一段时间，目前发现只要不改任何文件及配置的情况下，type-C可以正常加载，并且睡眠后雷电设备的信息仍然在。但只要我取消加载屏蔽独显的SSDT，那么我的雷电在PCI中就会消失，设备也无法被读取。因BIOS实则是一样的，所以独显屏蔽SSDT加载理论上不对机器产生负面影响，也就随它了。目前的问题就是开机速度略慢，且开机闪屏。不过我觉得应该都不会是大问题，感谢啦。 @gorquan

**Comment by gorquan** [comment]
> 好消息，我在重新安装我的macOS，当成主力用了一段时间，目前发现只要不改任何文件及配置的情况下，type-C可以正常加载，并且睡眠后雷电设备的信息仍然在。但只要我取消加载屏蔽独显的SSDT，那么我的雷电在PCI中就会消失，设备也无法被读取。因BIOS实则是一样的，所以独显屏蔽SSDT加载理论上不对机器产生负面影响，也就随它了。目前的问题就是开机速度略慢，且开机闪屏。不过我觉得应该都不会是大问题，感谢啦。 @gorquan

雷电是有雷电的SSDT的，这个SSDT只要你启用无论在什么状态下他都不会消失。不过这个雷电能不能用我就不确定，因为雷电设备实在是太贵了。TYPE-C沿用了黑果小兵的SSDT，从0.5.6开始我就没发现有问题。
开机慢我也无解，升级0.6.0之后开机确实比之前慢了。
开机闪屏其实是因为显存不对导致的，但是不影响使用，所以懒得修复了。应该把显存调回1536M可能会解决闪屏问题，但是也可能会带来其他奇葩问题。
屏蔽独显SSDT一定不能取消。。这个的目的是屏蔽独显让他不工作然后延长电池供电时间的，原来0.5.6的Opencore就是因为没有屏蔽独显，然后电池供电只能两个小时以下

**Comment by gorquan** [comment]
> 好消息，我在重新安装我的macOS，当成主力用了一段时间，目前发现只要不改任何文件及配置的情况下，type-C可以正常加载，并且睡眠后雷电设备的信息仍然在。但只要我取消加载屏蔽独显的SSDT，那么我的雷电在PCI中就会消失，设备也无法被读取。因BIOS实则是一样的，所以独显屏蔽SSDT加载理论上不对机器产生负面影响，也就随它了。目前的问题就是开机速度略慢，且开机闪屏。不过我觉得应该都不会是大问题，感谢啦。 @gorquan

准确来说，现在1080p版本的7590已经是白果状态，不会有什么问题了

**Comment by qianwuyu** [comment]
@gorquan 啊，其实是这样的。
- 首先，我取消独显的SSDT是因为我没有独显，哈哈哈。
- 其次，我的雷电SSDT即使加载了也会莫名其妙掉，也就是PCI信息显示不对。然后导致type-C设备也无法被读取。
- 再者，1536MB我改回来了据说改2048没有实际意义，哈哈哈，但还是闪。
- 有时候我重启声卡也会掉。

只能说我这台机器的兼容性太差了，可能是比较差的那批料做的XPS吧。而且我还是蛮想这台机器可以达到6-8小时续航的，毕竟电池够大，而且启动要20秒左右确实比较慢了，以前的黑果子都是十秒以内的，可能是有对比了，哈哈哈。

**Comment by qianwuyu** [comment]
@gorquan 另外，显卡信息在PCI那不太像以前我显卡驱动的样子，比如没有显示器信息，可能是config里配置了dpcd多链路的原因。的确，如你所说，已经可以被当成白苹果使用了，感谢你的适配，非常感谢，哈哈哈。

**Comment by qianwuyu** [comment]
Thank you for all who helped me before, I have found the problem why the device so strange in my machine. Just for the setting in bios which's involved thoundbolt 3. That option means that if you handle the operation control to the system or by native config info. All right, maybe it can be closed.

**Comment by HBUer** [comment]
Maybe I have already sole this Thounderbolt 3.
<img width="841" alt="截屏2021-03-23 下午3 25 24" src="https://user-images.githubusercontent.com/66448191/112109171-2f7a0280-8bec-11eb-837b-9eb434e84611.png">
It use a  kext.
[IOElectrify_10.15sdk.kext.zip](https://github.com/gorquan/OC-XPS-7590/files/6187669/IOElectrify_10.15sdk.kext.zip)
@gorquan

**Comment by gorquan** [comment]
> Maybe I have already sole this Thounderbolt 3.
> <img alt="截屏2021-03-23 下午3 25 24" width="841" src="https://user-images.githubusercontent.com/66448191/112109171-2f7a0280-8bec-11eb-837b-9eb434e84611.png">
> It use a kext.
> [IOElectrify_10.15sdk.kext.zip](https://github.com/gorquan/OC-XPS-7590/files/6187669/IOElectrify_10.15sdk.kext.zip)
> @gorquan

现在的OC也是可以识别到雷电3，但是能不能有雷电3的效果或支持雷电3的设备我也没办法测试。
